File: OpenGLES_Ch2_2/OpenGLES_Ch2_2/Shader.m
//
//  Shader.m
//  OpenGLES_Ch2_2
//
//  Created by boob on 2021/5/30.
//

#import <GLKit/GLKit.h>
#import "Shader.h"
@interface Shader()
{
    GLuint program;
}
@end

@implementation Shader

- (void)use
{
    glUseProgram(program);
}

- (void)setBool:(NSString *)name value:(BOOL)value
{
    glUniform1i(glGetUniformLocation(program, [name cStringUsingEncoding:NSUTF8StringEncoding]), value);
}

- (BOOL)getValue:(NSString *)key 
{ 
    return glGetAttribLocation(program, [key cStringUsingEncoding:NSUTF8StringEncoding]); ;
}
- (void)setInt:(NSString *)name value:(int)value
{ 
    glUniform1i(glGetUniformLocation(program, [name cStringUsingEncoding:NSUTF8StringEncoding]), value);
}

- (void)setFloat:(NSString *)name value:(float)value
{ 
    glUniform1f(glGetUniformLocation(program, [name cStringUsingEncoding:NSUTF8StringEncoding]), value);
}

- (BOOL)loadShaders:(NSString *)shadervshstr shaderfshstr:(NSString *)shaderfshstr
{
    GLuint vertShader, fragShader;
    NSString *vertShaderPathname, *fragShaderPathname;
    
    // Create shader program.
    program = glCreateProgram();
    
    // Create and compile vertex shader.
    vertShaderPathname = [[NSBundle mainBundle] pathForResource:@"Shader" ofType:@"vsh"];
    if (![self compileShader:&vertShader type:GL_VERTEX_SHADER file:vertShaderPathname shaderstr:shadervshstr]) {
        NSLog(@"Failed to compile vertex shader");
        return NO;
    }
    
    // Create and compile fragment shader.
    fragShaderPathname = [[NSBundle mainBundle] pathForResource:@"Shader" ofType:@"fsh"];
    if (![self compileShader:&fragShader type:GL_FRAGMENT_SHADER file:fragShaderPathname shaderstr:shaderfshstr]) {
        NSLog(@"Failed to compile fragment shader");
        return NO;
    }
    
    // Attach vertex shader to program.
    glAttachShader(program, vertShader);
    
    // Attach fragment shader to program.
    glAttachShader(program, fragShader);
    
    // Link program.
    if (![self linkProgram:program]) {
        NSLog(@"Failed to link program: %d", program);
        
        if (vertShader) {
            glDeleteShader(vertShader);
            vertShader = 0;
        }
        if (fragShader) {
            glDeleteShader(fragShader);
            fragShader = 0;
        }
        if (program) {
            glDeleteProgram(program);
            program = 0;
        }
        
        return NO;
    }
    // Release vertex and fragment shaders.
    if (vertShader) {
        glDetachShader(program, vertShader);
        glDeleteShader(vertShader);
    }
    if (fragShader) {
        glDetachShader(program, fragShader);
        glDeleteShader(fragShader);
    }
    
    return YES;
}
 
- (BOOL)compileShader:(GLuint *)shader type:(GLenum)type file:(NSString *)file shaderstr:(NSString *)shaderstr
{
    GLint status;
    const GLchar *source;
    
    if (shaderstr) {
        source = [shaderstr cStringUsingEncoding:NSUTF8StringEncoding];  
    }else {
        source = (GLchar *)[[NSString stringWithContentsOfFile:file encoding:NSUTF8StringEncoding error:nil] UTF8String];    
    }

    if (!source) {
        NSLog(@"Failed to load vertex shader");
        return NO;
    }
    
    *shader = glCreateShader(type);
    glShaderSource(*shader, 1, &source, NULL);
    glCompileShader(*shader);
    
#if defined(DEBUG)
    GLint logLength;
    glGetShaderiv(*shader, GL_INFO_LOG_LENGTH, &logLength);
    if (logLength > 0) {
        GLchar *log = (GLchar *)malloc(logLength);
        glGetShaderInfoLog(*shader, logLength, &logLength, log);
        NSLog(@"Shader compile log:\n%s", log);
        free(log);
    }
#endif
    
    glGetShaderiv(*shader, GL_COMPILE_STATUS, &status);
    if (status == 0) {
        glDeleteShader(*shader);
        return NO;
    }
    
    return YES;
}
 
- (BOOL)linkProgram:(GLuint)prog
{
    GLint status;
    glLinkProgram(prog);
    
#if defined(DEBUG)
    GLint logLength;
    glGetProgramiv(prog, GL_INFO_LOG_LENGTH, &logLength);
    if (logLength > 0) {
        GLchar *log = (GLchar *)malloc(logLength);
        glGetProgramInfoLog(prog, logLength, &logLength, log);
        NSLog(@"Program link log:\n%s", log);
        free(log);
    }
#endif
    
    glGetProgramiv(prog, GL_LINK_STATUS, &status);
    if (status == 0) {
        return NO;
    }
    
    return YES;
}
 
- (BOOL)validateProgram:(GLuint)prog
{
    GLint logLength, status;
    
    glValidateProgram(prog);
    glGetProgramiv(prog, GL_INFO_LOG_LENGTH, &logLength);
    if (logLength > 0) {
        GLchar *log = (GLchar *)malloc(logLength);
        glGetProgramInfoLog(prog, logLength, &logLength, log);
        NSLog(@"Program validate log:\n%s", log);
        free(log);
    }
    
    glGetProgramiv(prog, GL_VALIDATE_STATUS, &status);
    if (status == 0) {
        return NO;
    }
    
    return YES;
}

@end

File: learnopengl/glfw-glsl/glfw-hellotriangle/Shader.cpp
//
//  Shader.cpp
//  glfw-glsl
//
//  Created by boob on 2021/6/27.
//

#include "Shader.hpp"

Shader::Shader()
{
    
}
Shader::~Shader()
{
    
}
Shader::Shader(const GLchar* vertexPath, const GLchar* fragmentPath)
{
    // 1. 从文件路径中获取顶点/片段着色器
    std::string vertexCode;
    std::string fragmentCode;
    std::ifstream vShaderFile;
    std::ifstream fShaderFile;
    // 保证ifstream对象可以抛出异常：
    vShaderFile.exceptions(std::ifstream::badbit);
    fShaderFile.exceptions(std::ifstream::badbit);
    try
    {
        // 打开文件
        vShaderFile.open(vertexPath);
        fShaderFile.open(fragmentPath);
        std::stringstream vShaderStream, fShaderStream;
        // 读取文件的缓冲内容到流中
        vShaderStream << vShaderFile.rdbuf();
        fShaderStream << fShaderFile.rdbuf();
        // 关闭文件
        vShaderFile.close();
        fShaderFile.close();
        // 转换流至GLchar数组
        vertexCode = vShaderStream.str();
        fragmentCode = fShaderStream.str();
    }
    catch(std::ifstream::failure e)
    {
        std::cout << "ERROR::SHADER::FILE_NOT_SUCCESFULLY_READ" << std::endl;
    }
    const GLchar* vShaderCode = vertexCode.c_str();
    const GLchar* fShaderCode = fragmentCode.c_str();
//    [...]
    
    //MARK: create shader
    
    GLuint vertextShader;
    vertextShader = glCreateShader(GL_VERTEX_SHADER);
    glShaderSource(vertextShader,1,&vShaderCode, NULL);
    glCompileShader(vertextShader);
    
    GLint success;
    GLchar infoLog[512];
    glGetShaderiv(vertextShader,GL_COMPILE_STATUS,&success);
    if(!success) {
        glGetShaderInfoLog(vertextShader,512,NULL,infoLog);
        std::cout << "error: shader::vertext::compilation_failed\n" << infoLog << std::endl;
    }
    
    GLuint fragmentShader;
   fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
   glShaderSource(fragmentShader,1,&fShaderCode,NULL);
   glCompileShader(fragmentShader);
   
//   GLint success;
//   GLchar infoLog[512];
   glGetShaderiv(fragmentShader,GL_COMPILE_STATUS,&success);
   if(!success) {
       glGetShaderInfoLog(fragmentShader,512,NULL,infoLog);
       std::cout << "error: shader::fragment::compilation_failed\n" << infoLog << std::endl;
   }
    
    //MARK: attach && link
    
    GLuint shaderPrograme;
    shaderPrograme = glCreateProgram();
    Program = shaderPrograme;
    glAttachShader(shaderPrograme,vertextShader);
    glAttachShader(shaderPrograme,fragmentShader);
    glLinkProgram(shaderPrograme);
     
    glGetProgramiv(shaderPrograme,GL_LINK_STATUS,&success);
    if(!success) {
        glGetProgramInfoLog(shaderPrograme,512,NULL,infoLog);
        std::cout << "error::progreame::linkerror:\n"<< infoLog << std::endl;
    }
    
    glDeleteShader(vertextShader);
    glDeleteShader(fragmentShader);
//    return shaderPrograme;
     
}

void Shader::Use()
{
    glUseProgram(this->Program);
}
